### PR TITLE
Bank and cafe

### PR DESCRIPTION
**Pull request by Mar-Alex**
Tasks are completed.

### CODE REVIEW

File: src/main/java/school/lemon/changerequest/java/banking/impl/BankAccountImpl.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+public class BankAccountImpl implements BankAccount {
+
+    private static int numberOfAccounts = 0;
+
+    protected double balance;
+    protected double rate;
+    protected final int accountNumber;
+
+    public BankAccountImpl() {
+        this(0, 0.01);
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/BankAccountImpl.java]: Make `0.01` class-level constant.

File: src/main/java/school/lemon/changerequest/java/banking/impl/BankAccountImpl.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    public BankAccountImpl(double balance, double rate) throws IllegalArgumentException {
+        if (balance < 0 || rate < 0)
+            throw new IllegalArgumentException("Balance < 0 or(and) rate < 0");
+        this.balance = balance;
+        this.rate = rate;
+        accountNumber = numberOfAccounts++;
+    }
+
+    public BankAccountImpl(TransactionalBankAccountImpl obj) {
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/BankAccountImpl.java]: Why do you need this constructor? Parent classes should know nothing about their children in almost all cases.

File: src/main/java/school/lemon/changerequest/java/banking/impl/TransactionalBankAccountImpl.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package school.lemon.changerequest.java.banking.impl;
+
+import school.lemon.changerequest.java.banking.TransactionalBankAccount;
+
+public class TransactionalBankAccountImpl extends BankAccountImpl implements TransactionalBankAccount {
+
+    private BankAccountImpl latestCommit;
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/TransactionalBankAccountImpl.java]: Why do you need instance of the whole bank account here? If you need to save some information — use appropriate types for it. E.g. create additional internal class to represent last account state.

File: src/main/java/school/lemon/changerequest/java/banking/impl/TransactionalBankAccountImpl.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    private boolean autoCommit = true;
+
+
+    public TransactionalBankAccountImpl() {
+        this(0, 0);
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/TransactionalBankAccountImpl.java]: Why don't you simply call `super`?

File: src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java [hunk 5]
```diff
@@ -11,25 +11,60 @@
...
+            throw new NullPointerException("bankAccount == null");
+        this.bankAccount = bankAccount;
     }
 
     /**
      * Execute bunch of operations.
+     *
      * @param operations array of account operations
      */
     public void execute(BankOperation[] operations) {
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java:24]: Take a look at the docs:
```text
execute - executes array of operations. Do not guaranty that all operations are executed without errors and do not revert previous operations in case of errors.
```

What will happen to the account if second operation fails? You should be sure, that operations are applied immediately.

File: src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java [hunk 6]
```diff
@@ -11,25 +11,60 @@
...
      * In case of any exception all previous operations should be reverted and original exception should be thrown.
      * {@code autoCommit} should be enabled in any case.
+     *
      * @param operations array of account operations
      */
     public void executeInTransaction(BankOperation[] operations) throws IllegalArgumentException {
-        //TODO: implement me
+
+        try {
+            bankAccount.commit();
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java]: Why are you commiting changes here? Is it required from the docs?

File: src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java [hunk 7]
```diff
@@ -11,25 +11,60 @@
...
+            bankAccount.setAutoCommit(false);
+            execute(operations);
+        } catch (IllegalArgumentException e) {
+            bankAccount.revert();
+            throw e;
+        } finally {
+            bankAccount.setAutoCommit(true);
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java:48]: OK, what if all operations are applied? Why aren't they applied afterwards?

File: src/main/java/school/lemon/changerequest/java/cafe/domain/Cafe.java [hunk 8]
```diff
@@ -2,17 +2,39 @@
...
 public class Cafe {
 
     private final Random random = new Random();
 
-    public void serve(Client client) {
-        client.drinkCoffee(generateCup());
+    public void serve(Client client) throws RandomClientException {
+        Drink drink = null;
+        try {
+            drink = generateCup();
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/cafe/domain/Cafe.java]: `generateCup` doesn't throw any exception — move it out of `try-catch` block.

File: src/main/java/school/lemon/changerequest/java/cafe/domain/Cafe.java [hunk 9]
```diff
@@ -2,17 +2,39 @@
...
+            client.drinkCoffee(drink);
+        } catch (IncorrectDrinkException e) {
+            System.out.println(e.getMessage());
+            serve(client);
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/cafe/domain/Cafe.java]: You are going into recursion. It's not the best way to handle current problem. Can you propose any other method?

File: src/main/java/school/lemon/changerequest/java/cafe/domain/Client.java [hunk 10]
```diff
@@ -9,15 +11,16 @@
 
     private final Random random = new Random();
 
-    public void drinkCoffee(Drink drink) {
+    public void drinkCoffee(Drink drink) throws IncorrectDrinkException, LowTemperatureException,
+            HighTemperatureException, RandomClientException {
         if (drink.getDrinkType() != DrinkType.COFFEE) {
-            System.out.println("Tastes like a " + drink.getDrinkType());
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/cafe/domain/Client.java]: Why did you remove those output?

File: src/main/java/school/lemon/changerequest/java/cafe/domain/Client.java [hunk 11]
```diff
@@ -9,15 +11,16 @@
...
+            throw new IncorrectDrinkException("Tastes like a " + drink.getDrinkType());
         } else if (drink.getTemperature() < LOWEST_TEMPERATURE) {
-            System.out.println("Looks like it's too cold...");
+            throw new LowTemperatureException("Looks like it's too cold...", drink.getTemperature());
         } else if (drink.getTemperature() > HIGHEST_TEMPERATURE) {
-            System.out.println("Looks like it's too hot...");
+            throw new HighTemperatureException("Looks like it's too hot...", drink.getTemperature());
         } else if (random.nextInt(100) < 10) {
-            System.out.println("I don't like your Cafe!");
+            throw new RandomClientException("I don't like your Cafe!");
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/cafe/domain/Client.java]: It's definitely not `random` exception. It's an exception that says smth like: 'Hey, I don't like you cafe' — maybe smth is wrong with serving or anything else, but please create meaningful exception name.

File: src/main/java/school/lemon/changerequest/java/cafe/exceptions/LowTemperatureException.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package school.lemon.changerequest.java.cafe.exceptions;
+
+public class LowTemperatureException extends ClientException {
```
Comment (yuri-sergiichuk) [inline src/main/java/school/lemon/changerequest/java/cafe/exceptions/LowTemperatureException.java]: Looks like `Low` and `High` temperature exceptions should be from the same hierarchy... You have duplicated code.

**Review by yuri-sergiichuk — CHANGES_REQUESTED** [review]
@Mar-Alex please take a look at the comments below.

File: src/main/java/school/lemon/changerequest/java/banking/impl/TransactionalBankAccountImpl.java (hunk 4 above)
Comment (Mar-Alex) [reply src/main/java/school/lemon/changerequest/java/banking/impl/TransactionalBankAccountImpl.java]: Because we must create first <code>latestCommit</code> in constructor

File: src/main/java/school/lemon/changerequest/java/cafe/domain/Client.java (hunk 10 above)
Comment (Mar-Alex) [reply src/main/java/school/lemon/changerequest/java/cafe/domain/Client.java]: In original <code>main</code> method block <code>catch</code> was responsible for outputing information about exception of fourth type (unsatisfied client). So, I decided that other types of exceptions must be printed by <code>catch</code> too.

**Comment by Mar-Alex** [comment]
It looks like a misprint:
1)In <code>README</code> "Should disable autoCommit after execution."
2)In comment for method <code>executeInTransaction</code> of class <code>AccountTransactionManager</code> " {@code autoCommit} should be enabled in any case"
Which variant is right?

**Comment by xSAVIKx** [comment]
@Mar-Alex there is a misprint in `README`. I'll fix it.

File: src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java (hunk 6 above)
Comment (Mar-Alex) [reply src/main/java/school/lemon/changerequest/java/banking/impl/AccountTransactionManager.java]: Can I add new methods to the interface ? I need <code>get</code> and <code>set</code> methods for <code>latestCommit</code>. 
My idea is to copy <code>latestCommit</code>, then commit changes, try to do operations and revert them in case of an exception and afterall recover the initial state of <code>latestCommit</code>

**Comment by Mar-Alex** [comment]
I am not sure how to fix problem with advance commit in <code>executeInTransaction</code>. I see two ways:
1)Add <code>setLatestCommit</code> and <code>getLatestCommit</code> to the interface, but it damages encapsulation.
2)Add these methods to the implementation and use type conversion, but in such case <code>executeInTransaction</code> can`t be used for other classes that implements <code>TransactionalBankAccount</code> interface.
Could you advise me a better solution?

**Comment by xSAVIKx** [comment]
@Mar-Alex Take a look at reference implementation [here](https://github.com/didva/exceptions.hw1).
